### PR TITLE
Add exponential retry on fetch

### PR DESCRIPTION
**Pull request by go-dima**
### What does this PR do?
- Add a retry for 500 server error  using exponential backoff mechanism

### CODE REVIEW

**Comment by go-dima** [comment]
PR fails due to missing dependency -backoff module.
What is the correct way to add this in the project? The options I found weer requirements.txt (didn't work) or in the dockerfile itself.

**Comment by rpothier** [comment]
Hi @go-dima 
Looking at the error message.
ERROR: Found 1 import issue(s) on python 2.7 which need to be resolved:
ERROR: plugins/lookup/conjur_variable.py:100:0: traceback: ImportError: No module named backoff
See documentation for help: https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/import.html

The issue seems to be that Ansible is requiring the import in a try/except block.
I just added comments on the review.

File: plugins/lookup/conjur_variable.py
Comment (rpothier) [inline plugins/lookup/conjur_variable.py:100]: This should be a try/except block as described in [https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/import.html](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/import.html)
```
try:
    import backoff
except ImportError as imp_exc:
    BACKOFF_IMPORT_ERROR = imp_exc
else:
    BACKOFF_IMPORT_ERROR = None
 ```

File: requirements.txt
Comment (rpothier) [inline requirements.txt:2]: do we still need this?

File: plugins/lookup/conjur_variable.py
Comment (go-dima) [reply plugins/lookup/conjur_variable.py:100]: I'll take a look. Thanks.

File: requirements.txt
Comment (go-dima) [reply requirements.txt:2]: It's part of the attempt to handle the import error, will be removed

File: plugins/lookup/conjur_variable.py
Comment (go-dima) [reply plugins/lookup/conjur_variable.py]: @rpothier, Same result. It looks like the module isn't found, although it is supported officially.

Comment (rpothier) [reply plugins/lookup/conjur_variable.py]: I'm looking into it also. When you say it's officially supported, do you mean by python? or Ansible?

Comment (rpothier) [reply plugins/lookup/conjur_variable.py]: Hi @go-dima I did some research with Ansible and it looks like they do not support non standard libraries.
Ansible has `ansible.module_utils.api.retry` can you use that instead?

Comment (go-dima) [reply plugins/lookup/conjur_variable.py]: Looks like ansible retry doesn't fit - it expects an exception to identify failure for retry. I'll keep looking for a suitable solution

Comment (go-dima) [reply plugins/lookup/conjur_variable.py]: @rpothier 
I decided to implement it myself, care to tae a look? 
I also need an approval to run workflows now :)

**Comment by go-dima** [comment]
Looks like I have yml issues, although my change didn't involve such.
I'll be glad to get some help on the matter.

**Comment by rpothier** [comment]
Hi @go-dima, I tried rerunning it, but same results. I'll try a baseline run.

**Comment by rpothier** [comment]
@go-dima I have posted #61 with a fix for the sanity errors.

**Comment by go-dima** [comment]
Thanks for the update! I'll wait for #61 and update this one.

**Comment by go-dima** [comment]
@rpothier can you re-trigger the build please? (I don't have permissions)

**Comment by rpothier** [comment]
@go-dima Did you rebase with the latest?

**Review by rpothier — COMMENTED** [review]
Can you summarize the testing that was done to validate this?

**Comment by rpothier** [comment]
@go-dima Woohoo! tests all pass. Changes look good.
Please squash the commits and can you describe the testing that was done?

**Comment by go-dima** [comment]
@rpothier 
Latest was merged 👍 

Regarding the tests:
We work with a docker worker to run our plays, so I changed the image to include the plugin from my branch. I triggered a workflow that uses the plugin multiple times in different places, to make sure the change doesn't break the successful flow.

I don't have a proper way to make sure that the retry happens during the plugin (since I cannot force 500 reply), but I did test the code in a standalone scenario with a simulated 500 error.

Regarding the squash - I plan to perform the merge itself with 'squash and merge' option.

**Review by rpothier — APPROVED** [review]
LGTM

**Comment by go-dima** [comment]
@rpothier 
I don't have write access (no surprising actually...)
Can you merge please? 🥇

**Comment by rpothier** [comment]
@go-dima Can you sign off your commit ? https://github.com/cyberark/community/blob/master/CONTRIBUTING.md#signing-off-a-commit

**Comment by go-dima** [comment]
@rpothier  Done

**Review by rpothier — APPROVED** [review]
LGTM

**Pull request merged**
